cpp/src/lance/arrow/dataset.cc

```diff
@@ -316,6 +317,39 @@ ::arrow::Result<std::shared_ptr<UpdaterBuilder>> LanceDataset::NewUpdate(
                                           std::move(new_field));
 }
 
+::arrow::Result<std::shared_ptr<LanceDataset>> LanceDataset::AddColumn(
+    const std::shared_ptr<::arrow::Field>& field, ::arrow::compute::Expression expression) {
+  auto refs = ::arrow::compute::FieldsInExpression(expression);
+  auto columns = refs | views::transform([](auto& ref) { return ref.ToString(); }) |
+                 to<std::vector<std::string>>;
+  ARROW_ASSIGN_OR_RAISE(auto builder, NewUpdate(field));
+  // Add column
+  // builder->Project(columns);
+  ARROW_ASSIGN_OR_RAISE(auto updater, builder->Finish());
+
+  while (true) {
+    ARROW_ASSIGN_OR_RAISE(auto batch, updater->Next());
+    if (!batch) {
+      break;
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto datum,
+                          ::arrow::compute::ExecuteScalarExpression(expression, *schema(), batch));
+    fmt::print("Datum: {}\n", datum.ToString());
+    std::shared_ptr<::arrow::Array> arr;
+    if (datum.is_scalar()) {
+      ARROW_ASSIGN_OR_RAISE(arr, CreateArray(datum.scalar(), batch->num_rows()));
+    } else if (datum.is_chunked_array()) {
+      auto chunked_arr = datum.chunked_array();
+      ARROW_ASSIGN_OR_RAISE(arr, ::arrow::Concatenate(chunked_arr->chunks()));
```

Comment:
ExtensionArray's cannot be concatenated currently - tho compute expressions won't either so ExtensionArray's probably won't make it past ExecuteScalarExpression?

Reply:
Let's add a test as follow up? But also, there is no function / kernel is available for extension types right now, this method might fail earlier.